### PR TITLE
TIP-842: Extract attribute model to Akeneo\Pim\Structure

### PR DESCRIPTION
**Pull request by arnolanglade**
**Description (for Contributor and Core Developer)**

We are creating the new bounded context **Pim**. Normally, it should not be a BC but we won't have time to properly well organize it. This temporary BC contains two name spaces **Enrichment** and **"Structure"** which will become BC one day. **Structure** is about defining the product foundation like family, variant family, family requirement, attribute and attribute option. We decided to create lot of small PRs to do this task. First we will move all models (listed previously) in different small PRs. So, this is the first one, we moved Attribute model and its interface in the new name space.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | yes
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Comment by arnolanglade** [comment]
The change log is fixed in 01505b73916d8ea29a85e717051a2c80629167b6

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/BooleanFilter.php [hunk 1]
```diff
@@ -99,8 +99,8 @@ public function addAttributeFilter(
     /**
      * Check if the value is valid
      *
-     * @param AttributeInterface $attribute
-     * @param mixed              $value
+     * @param \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/BooleanFilter.php]: revert

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/MetricFilter.php [hunk 2]
```diff
@@ -231,8 +231,8 @@ protected function checkValue(AttributeInterface $attribute, $data)
     /**
      * Converts the given value to the base_unit configured in the family.
      *
-     * @param AttributeInterface $attribute
-     * @param array              $data
+     * @param \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/MetricFilter.php]: revert

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/NumberFilter.php [hunk 3]
```diff
@@ -139,8 +139,8 @@ public function addAttributeFilter(
     /**
      * Checks that the value is a number.
      *
-     * @param AttributeInterface $attribute
-     * @param mixed              $value
+     * @param \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/NumberFilter.php]: revert

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/PriceFilter.php [hunk 4]
```diff
@@ -243,8 +243,8 @@ protected function checkAmount(AttributeInterface $attribute, $data)
     }
 
     /**
-     * @param AttributeInterface $attribute
-     * @param mixed              $data
+     * @param \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/Attribute/PriceFilter.php]: revert

File: src/Pim/Bundle/CatalogBundle/Entity/Family.php [hunk 5]
```diff
@@ -39,10 +39,10 @@ class Family implements FamilyInterface
     /** @var Collection */
     protected $translations;
 
-    /** @var AttributeInterface */
+    /** @var \Akeneo\Pim\Structure\Component\Model\AttributeInterface */
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Entity/Family.php]: revert

File: src/Pim/Bundle/CatalogBundle/Entity/Family.php [hunk 6]
```diff
@@ -39,10 +39,10 @@ class Family implements FamilyInterface
...
     protected $attributeAsLabel;
 
-    /** @var AttributeInterface */
+    /** @var \Akeneo\Pim\Structure\Component\Model\AttributeInterface */
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Entity/Family.php]: revert

File: src/Pim/Bundle/CatalogBundle/tests/integration/Completeness/AbstractCompletenessTestCase.php [hunk 7]
```diff
@@ -70,7 +70,7 @@ protected function assertCompletenessesCount(ProductInterface $product, $expecte
      * @param bool   $scopable
      * @param array  $localesSpecific
      *
-     * @return AttributeInterface
+     * @return \Akeneo\Pim\Structure\Component\Model\AttributeInterface
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Completeness/AbstractCompletenessTestCase.php]: revert

File: src/Pim/Bundle/CatalogBundle/tests/integration/Validation/Attribute/AbstractAttributeTestCase.php [hunk 8]
```diff
@@ -432,16 +432,16 @@ protected function createAttribute()
     }
 
     /**
-     * @param AttributeInterface $attribute
-     * @param array              $data
+     * @param \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Validation/Attribute/AbstractAttributeTestCase.php]: revert

File: src/Pim/Bundle/CatalogBundle/tests/integration/Validation/Attribute/AbstractAttributeTestCase.php [hunk 9]
```diff
@@ -432,16 +432,16 @@ protected function createAttribute()
...
+     * @param array                                                    $data
      */
     protected function updateAttribute(AttributeInterface $attribute, array $data)
     {
         $this->get('pim_catalog.updater.attribute')->update($attribute, $data);
     }
 
     /**
-     * @param AttributeInterface $attribute
+     * @param \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Validation/Attribute/AbstractAttributeTestCase.php]: all `@param \Akeneo\Pim\Structure\Component\Model\AttributeInterface`

File: src/Pim/Component/Catalog/spec/Model/ProductModelSpec.php [hunk 10]
```diff
@@ -47,7 +47,7 @@ function it_is_a_category_aware_entity()
     function it_adds_a_value(
         ValueCollectionInterface $values,
         ValueInterface $value,
-        AttributeInterface $attribute
+        \Akeneo\Pim\Structure\Component\Model\AttributeInterface $attribute
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/spec/Model/ProductModelSpec.php]: all `\Akeneo\Pim\Structure\Component\Model\AttributeInterface`

File: src/Pim/Bundle/CatalogBundle/spec/Doctrine/Common/Saver/AttributeSaverSpec.php [hunk 11]
```diff
@@ -87,7 +87,7 @@ function it_throws_exception_when_save_anything_else_than_an_attribute()
             ->shouldThrow(
                 new \InvalidArgumentException(
                     sprintf(
-                        'Expects a "Pim\Component\Catalog\Model\AttributeInterface", "%s" provided.',
+                        'Expects a "Akeneo\Pim\Structure\Component\Model\AttributeInterface", "%s" provided.',
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/spec/Doctrine/Common/Saver/AttributeSaverSpec.php]: Would have been better to use `AttributeInterface::class` here.

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php [hunk 12]
```diff
@@ -410,7 +410,7 @@ protected function validateLocalizedData($data)
 
     /**
      * @param AttributeInterface $attribute
-     * @param array              $data
+     * @param array                                                    $data
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php]: useless space

Comment (damien-carcel) [reply src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php]: There are other case like this one in following classes.

File: src/Pim/Component/Catalog/Validator/Constraints/CurrencyValidator.php [hunk 13]
```diff
@@ -38,7 +38,7 @@ public function __construct(CurrencyRepositoryInterface $currencyRepository)
      * Validate currency of a price
      *
      * @param AttributeInterface|MetricInterface|ValueInterface $object
-     * @param Constraint                                        $constraint
+     * @param Constraint                                                                              $constraint
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Validator/Constraints/CurrencyValidator.php]: useless space too, many other after this one too.

**Pull request merged**